Adding browserName capability for mobile

vikram1711 opened the pull request:
Re-purposing the browserName capability to set the mobile browser name, the same way it is being used by desktop test

	modified:   appium-provider/src/main/java/com/paypal/selion/appium/internal/platform/grid/browsercapabilities/AppiumCapabilitiesBuilder.java
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/android/sample/AppiumAndroidBrowserSauceCloudTest.java
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/android/sample/AppiumAndroidBrowserTest.java
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/ios/sample/AppiumIOSSafariSauceCloudTest.java
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/ios/sample/AppiumIOSTest.java
	new file:   appium-provider/src/test/resources/apps/PageObjectsDemoApp.apk
	modified:   client/src/main/java/com/paypal/selion/annotations/MobileTest.java
	modified:   client/src/main/java/com/paypal/selion/internal/platform/grid/MobileTestSession.java

- [x ] By placing an `X` in the preceding checkbox, I verify that I am a PayPal employee or
I have signed the [Contributor License Agreement](https://github.com/paypal/SeLion#contributing)

---

Commit:
In latest version of appium- appium expect browserName capability to be set explicitly. Using the existing capability browserName to be used while creating mobile sessions.
	modified:   appium-provider/src/main/java/com/paypal/selion/appium/internal/platform/grid/browsercapabilities/AppiumCapabilitiesBuilder.java
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/android/sample/AppiumAndroidBrowserSauceCloudTest.java
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/android/sample/AppiumAndroidBrowserTest.java
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/ios/sample/AppiumIOSSafariSauceCloudTest.java
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/ios/sample/AppiumIOSTest.java
	new file:   appium-provider/src/test/resources/apps/PageObjectsDemoApp.apk
	modified:   client/src/main/java/com/paypal/selion/annotations/MobileTest.java
	modified:   client/src/main/java/com/paypal/selion/internal/platform/grid/MobileTestSession.java

mach6 commented on client/src/main/java/com/paypal/selion/internal/platform/grid/MobileTestSession.java:
```diff
@@ -212,14 +224,21 @@ public void initializeTestSession(InvokedMethodInformation method) {
+        //App run and browser run should be mutually exclusive. If appName or appPath is provided disregard
+        //the browser parameter
+        
+        if(StringUtils.isNotBlank(appName) || StringUtils.isNotBlank(appPath)){
```
slight nit pick here... 
please fix the whitespace so that it is consistent with the rest of this file.

mach6 commented:
please update the date range on the copyright header for each `.java` file modified.

mach6 commented on client/src/main/java/com/paypal/selion/internal/platform/grid/MobileTestSession.java:
```diff
@@ -212,14 +224,21 @@ public void initializeTestSession(InvokedMethodInformation method) {
-                "Either you have provided both appPath and appName or you have specified nothing. Please specify either "
-                + "appPath or appName");
+                StringUtils.isNotBlank(appName) ^ StringUtils.isNotBlank(appPath) ^ StringUtils.isNotBlank(browserName),
+                "Either you have provided appPath, appName and browserName all three of them or you have specified nothing. Please specify either "
```
I suggest the text change to the following...
`Either you have provided appPath, appName, and browserName all at once or you have specified nothing. Please specify one of appPath, appName, or browserName`

mach6 commented on appium-provider/src/test/java/com/paypal/selion/appium/ios/sample/AppiumIOSSafariSauceCloudTest.java:
```diff
@@ -31,7 +31,8 @@
     @Test
-    @MobileTest(appName = "safari", device = "iphone:8.1", deviceType = "iPhone Simulator", additionalCapabilities = {
+    @MobileTest
+    (browserName = "safari", device = "iphone:10.2", deviceType = "iPhone SE", additionalCapabilities = {
```
please don't introduce this line break

mach6 commented:
@vikram1711 is the file `appium-provider/src/test/resources/apps/PageObjectsDemoApp.apk` introduced to fix an unrelated (to this change set) test?

mach6 commented:
@vikram1711 all-in-all this looks good.. I just have the few suggestions/feekback. Cheers!

vikram1711 commented:
I ran native app test just to make sure, this change works seamlessly for native app as well. While running the test I figured out -provider/src/test/resources/apps/PageObjectsDemoApp.apk is missing from specified location hence added it, so that these test don't fail in future.

Commit:
Making changes based on review comments
	modified:   appium-provider/src/test/java/com/paypal/selion/appium/ios/sample/AppiumIOSSafariSauceCloudTest.java
	modified:   client/src/main/java/com/paypal/selion/annotations/MobileTest.java
	modified:   client/src/main/java/com/paypal/selion/internal/platform/grid/MobileTestSession.java

sebady commented on client/src/main/java/com/paypal/selion/annotations/MobileTest.java:
```diff
@@ -124,5 +124,11 @@
+     * 
+     */
 
+    String browserName() default "";
```
I'm not sure if this is ideal to have an additional field for browserName in the mobile use case.  The appName field is intended to work even for browsers values (e.g.  "safari").

vikram1711 replied:
@sebady and @mach6 -As far as I know appium would expect a separate capability "browserName" in order to create a browser based session and "app" to create an app based session. 

If you try to set app=safari, appium will complain that "app name does not have extension .app" or in case of chrome ".apk", that's why we need this capability to set the browserName separately.

Other observation- testing infrastructure providers such as sauce they also expect a browserName capability in order to create a browser based session - https://wiki.saucelabs.com/display/DOCS/Platform+Configurator#/

mach6 replied:
What is even more annoying is in the case of an app or hybrid test, they still expect a capability `browserName: ""`, I believe.

mach6 commented:
@vikram1711 thanks for this. merged it
